### PR TITLE
meta/autoid: fix ID allocator when transaction failed to commit.

### PR DESCRIPTION
**Pull request by coocood**
During the execution of transaction, the allocator update its field `base` and `end`,
But didn't set it back when transaction fails.

This PR fix this issue.

### CODE REVIEW

**Comment by coocood** [comment]
@disksing @shenli PTAL

File: parser/goyacc/main.go [hunk 1]
```diff
@@ -388,7 +388,7 @@ type %[1]sXError struct {
 		case "$default":
 			nm = *oPref + "Default"
 		case "$end":
-			nm = *oPref + "EofCode"
+			nm = *oPref + "EOFCode"
```
Comment (shenli) [inline parser/goyacc/main.go]: Why change this?

Comment (coocood) [reply parser/goyacc/main.go]: golint reports error

**Comment by zimulala** [comment]
LGTM

**Comment by disksing** [comment]
How about `var base, end int64` before `RunInNewTxn`, then in the txn set `base` and `end`, and update `alloc.base` and `alloc.end` only the txn commit success?

**Comment by coocood** [comment]
@zimulala @disksing @shenli PTAL

File: meta/autoid/autoid.go [hunk 2]
```diff
@@ -56,47 +57,61 @@ func GetStep() int64 {
...
-			return errors.Trace(err)
+		_, err1 = m.GenAutoTableID(alloc.dbID, tableID, newEnd-currentEnd)
+		if err1 != nil {
+			return errors.Trace(err1)
 		}
-
-		alloc.end = end
-		alloc.base = newBase
 		if !allocIDs {
 			alloc.base = alloc.end
```
Comment (shenli) [inline meta/autoid/autoid.go]: update alloc.end here?

File: meta/autoid/autoid.go [hunk 3]
```diff
@@ -56,51 +57,60 @@ func GetStep() int64 {
...
+			return errors.Trace(err1)
 		}
-		end, err = m.GenAutoTableID(alloc.dbID, tableID, newStep)
-		if err != nil {
-			return errors.Trace(err)
-		}
-
-		alloc.end = end
-		alloc.base = newBase
 		if !allocIDs {
```
Comment (zimulala) [inline meta/autoid/autoid.go]: Can we remove this code now?

**Comment by zimulala** [comment]
LGTM

**Pull request merged**